### ISSUE TITLE
[Darc] Use JSON as the output format

### ISSUE THREAD

**Issue by jcagme** [dotnet/arcade#2099]
Currently Darc outputs the results of some commands in YAML. There has been several request, including build from source, to change the output to JSON.

FYI: @mmitche @chcosta @crummel

**Comment by dagood** [comment]
Why JSON in particular?

While implementing ProdCon v1, we decided to use XML for data like this unless there's a specific reason to use something else. The v2 dependency data is currently in XML. It also lets us stick with the standard library (JSON isn't quite in CoreFX yet). https://github.com/dagood/core-eng/pull/2#discussion_r155118300 (although I think it also came up in a meeting I can't find a record of).

**Comment by jcagme** [comment]
@chcosta and I talked to @crummel and he mentioned JSON was better so probably he can outline the reason on why JSON specifically.

**Comment by crummel** [comment]
JSON had gotten brought up by @mmitche too as something you guys were looking at already, XML would also be just fine for us.  It's a good point that we wouldn't have to bring in Newtonsoft.Json then, although this tooling won't be going into the tarball so we don't have to worry about prebuilts as much.

**Comment by RussKie** [comment]
.NET Core now support JSON natively, so I'd go with it because JSON is less verbose than xml.

**Comment by riarenas** [comment]
We have a volunteer for this (Thanks Igor!) but this is more of an improvement and not needed for .NET 5.

**Comment by mmitche** [comment]
@RussKie were you actively working on this? If not I'm going to implement it for gather-drops manifest.

**Comment by mmitche** [comment]
https://github.com/dotnet/arcade-services/pull/957

First bit of this, for the manifest generation.

**Comment by mmitche** [comment]
The manifest bit is now in rollout. I took a look through a lot of the rest of the uses of YAML in the code. It's a bit of a mixed bag. There were lots of places where non-YAML output was used as part of status messages, error messages, etc. Things like:
```
Triggering subscription blah
```
These could probably stay the same. But there are also a number of places where I wanted shorthand descriptions of things like default channels, existing channels, subscriptions where the output is very useful and more compact than yaml but most decidedly NOT yaml:

```
https://github.com/dotnet/aspnetcore-tooling (.NET Core 5 Dev) ==> 'https://github.com/dotnet/aspnetcore' ('master')
  - Id: b4589d4d-ab29-4a8a-d45b-08d76e1d3434
  - Update Frequency: EveryBuild
  - Enabled: True
  - Batchable: True
  - Merge Policies:
    Standard
```

A lot of times these non-yaml bits, including status messages and confirmation prompts, are mixed in with yaml. Simply converting the yaml to json will still leave non-yaml bits (`https://github.com/dotnet/aspnetcore-tooling (.NET Core 5 Dev) ==> 'https://github.com/dotnet/aspnetcore' ('master')`) as well as intermingled intermingled status messages, confirmation prompts, etc.. Even if the non-status message parts are changed to be json, the intermingled status will mess up any attempt to parse the tool output as json.

Perhaps the right thing to do is to only convert the pieces that are truly stateless and don't print status messages or need confirmation prompts (get-subscriptiosn comes to mind)

**Comment by mmitche** [comment]
https://github.com/dotnet/arcade/issues/4816 is part of this, in rollout now.

**Comment by mmitche** [comment]
https://github.com/dotnet/arcade/issues/4816 is part of this, in rollout now.

**Comment by mmitche** [comment]
https://github.com/dotnet/arcade-services/pull/960

**Comment by RussKie** [comment]
In my little app I had to write a parser and split `https://github.com/dotnet/aspnetcore-tooling (.NET Core 5 Dev) ==> 'https://github.com/dotnet/aspnetcore' ('master')` as follows:

```cs
Channel:       .NET Core 5 Dev
Source:        https://github.com/dotnet/aspnetcore-tooling
Target:        https://github.com/dotnet/aspnetcore
Target branch: master
```

I think JSON could do the same. Not sure about YAML, but probably too

**Comment by mmitche** [comment]
@Russkie I would keep the original yamlish format as i think it's better for human understanding, but make the json format basically the version of what you've shown above.

Maybe we should rebrand the "yaml" format as "text" instead to avoid confusion.

**Comment by dagood** [comment]
Consider fixing the yaml but keep the custom key?

```yaml
"https://github.com/dotnet/runtime (.NET Core 5 Dev) ==> 'https://github.com/dotnet/diagnostics' ('master')":
- Id: 644ad545-d4a4-4952-d433-08d76e1d3434
- Source Repo: https://github.com/dotnet/runtime
- Source Channel: .NET Core 5 Dev
- Target Repo: https://github.com/dotnet/diagnostics
- Target Branch: master
- Update Frequency: EveryDay
- Enabled: True
- Batchable: False
- Merge Policies:
    Standard
```

**Comment by dagood** [comment]
But I do think it'd be interesting to drop yaml support, make json official for interop, and make the output very readable.

**Comment by mmitche** [comment]
You mean two modes:
- Json for machine
- Non-yamly text stuff for humans?

Or do you mean json everywhere, but very readable output?

**Comment by dagood** [comment]
Yeah, sorry, meant what you were saying with changing "yaml" to a default "text" output that is very readable. Wanted to emphasize that anything that is yaml-y rather than readable should be replaced.

**Comment by RussKie** [comment]
Can we drop yaml? Is used somewhere? 
Json is generally quite readable (or can be easily formatted to be readable), and we have standard parsers for it.

**Comment by mmitche** [comment]
@RussKie a lot of the outputs are not yaml but specifically designed to be more compact, where json is going to be too verbose: For example:

```
PS D:\enlistments\extensions> darc trigger-subscriptions --target-repo aspnetcore --channel 5
Will trigger the following 3 subscriptions...
  https://github.com/dotnet/aspnetcore-tooling (.NET Core 5 Dev) ==> 'https://github.com/dotnet/aspnetcore' ('master')
  https://github.com/dotnet/efcore (.NET Core 5 Dev) ==> 'https://github.com/dotnet/aspnetcore' ('master')
  https://github.com/dotnet/extensions (.NET Core 5 Dev) ==> 'https://github.com/dotnet/aspnetcore-tooling' ('master')
```

For get-default-channels:

```
(772)  https://dev.azure.com/dnceng/internal/_git/aspnet-Blazor @ internal/release/3.0 -> .NET Core 3.0 Internal Servicing
(792)  https://dev.azure.com/dnceng/internal/_git/aspnet-Blazor @ internal/release/3.1 -> .NET Core 3.1 Internal Servicing
(769)  https://dev.azure.com/dnceng/internal/_git/aspnet-websdk @ internal/release/3.0.1xx -> .NET Core SDK 3.0.1xx Internal
(811)  https://dev.azure.com/dnceng/internal/_git/aspnet-websdk @ internal/release/3.1.2xx -> .NET Core SDK 3.1.2xx Internal
(796)  https://dev.azure.com/dnceng/internal/_git/aspnet-websdk @ internal/release/3.1.1xx -> .NET Core SDK 3.1.1xx Internal
(896)  https://dev.azure.com/dnceng/internal/_git/dotnet-arcade @ TestSDK -> General Testing
```
We could make this json, but json is for sure less readable than this format..which is really just deisgned for human consumption.

**Comment by RussKie** [comment]
Thank you for the clarifications. 
Then I agree with the idea of not calling it yaml, it is a yaml-like human readable text version.
